src/main/telemetry/smartport.c

```diff
@@ -380,29 +388,114 @@ static void resetMspPacket(mspPacket_t* packet)
     packet->result  = 0;
 }
 
-static void processMspPacket(mspPacket_t* packet)
+static void initSmartPortMspReply(int16_t cmd)
 {
     smartPortMspReply.buf.ptr    = smartPortMspTxBuffer;
     smartPortMspReply.buf.end    = ARRAYEND(smartPortMspTxBuffer);
 
-    smartPortMspReply.cmd    = -1;
+    smartPortMspReply.cmd    = cmd;
     smartPortMspReply.result = 0;
+}
 
-    const mspResult_e status = mspFcProcessCommand(packet, &smartPortMspReply, NULL);
+static void processMspPacket(mspPacket_t* packet)
+{
+    initSmartPortMspReply(0);
+
+    mspFcProcessCommand(packet, &smartPortMspReply, NULL);
+
+    // change streambuf direction
+    sbufSwitchToReader(&smartPortMspReply.buf, smartPortMspTxBuffer);
+    smartPortMspReplyPending = true;
+}
+
+/**
+ * Request frame format:
+ * - Header: 1 byte
+ *   - Reserved: 2 bits (future use)
+ *   - Error-flag: 1 bit
+ *   - Start-flag: 1 bit
+ *   - CSeq: 4 bits
+ *
+ * - MSP payload:
+ *   - if Error-flag == 0:
+ *     - size: 1 byte
+ *     - payload
+ *     - CRC (request type included)
+ *   - if Error-flag == 1:
+ *     - Error: 1 Byte
+ *       - 0: Version mismatch (type=0)
+ *       - 1: Sequence number error
+ *       - 2: MSP error
+ *     - CRC (request type included)
+ */
+bool smartPortSendMspReply()
+{
+    static uint8_t checksum = 0;
+    static uint8_t seq = 0;
 
-    if (status != MSP_RESULT_NO_REPLY) {
-        // change streambuf direction
-        sbufSwitchToReader(&smartPortMspReply.buf, smartPortMspTxBuffer);
+    uint8_t packet[SMARTPORT_PAYLOAD_SIZE];
+    uint8_t* p = packet;
+    uint8_t* end = p + SMARTPORT_PAYLOAD_SIZE;
+    
+    sbuf_t* txBuf = &smartPortMspReply.buf;
 
-        smartPortMspReplyPending = true;
+    // detect first reply packet
+    if (txBuf->ptr == smartPortMspTxBuffer) {
+
+        uint8_t size = sbufBytesRemaining(txBuf);
+
+        // header
+        
+        *p = SMARTPORT_MSP_START_FLAG | (seq++ & SMARTPORT_MSP_SEQ_MASK);
+
+        if (smartPortMspReply.result == MSP_RESULT_ERROR) {
+            *p++ |= SMARTPORT_MSP_ERROR_FLAG;
+            *p++ = SMARTPORT_MSP_ERROR;
+            smartPortSendPackageEx(FSSP_MSPS_FRAME,packet);
```

Comment:
random data are sent after SMARTPORT_MSP_ERROR, it is better to clear them (and the space may be used in future)